appcues/src/main/java/com/appcues/statemachine/StateMachine.kt

```diff
@@ -90,59 +96,55 @@ internal class StateMachine(
 
-            if (transition.emitStateChange) {
                 // emit state change to all listeners via flow
-                _stateFlow.emit(state)
+                if (emitStateChange) _stateFlow.emit(state)
             }
+
+            sideEffect.launch()
         }
+    }
 
-        return if (sideEffect != null) {
-            when (sideEffect) {
-                is ContinuationEffect -> {
-                    // recursive call on continuations to get the final return value
-                    handleActionInternal(sideEffect.action)
-                }
+    private suspend fun SideEffect?.launch(): ResultOf<State, Error> {
+        return when (this) {
+            is ContinuationEffect -> {
+                // recursive call on continuations to get the final return value
+                handleActionInternal(action)
+            }
 
-                is ReportErrorEffect -> {
-                    _errorFlow.emit(sideEffect.error)
-                    // return a failure if this call to `handleAction` ended with a reported error
-                    Failure(sideEffect.error)
-                }
+            is ReportErrorEffect -> {
+                _errorFlow.emit(error)
+                // return a failure if this call to `handleAction` ended with a reported error
+                Failure(error)
+            }
 
-                is PresentContainerEffect -> {
-                    // this will execute the presentation, asynchronously, and return
-                    // the subsequent action to take - either continue to Rendering, or report Error
-                    handleActionInternal(sideEffect.presentContainer(actionProcessor))
-                }
+            is PresentContainerEffect -> {
+                // this will execute the presentation, asynchronously, and return
+                // the subsequent action to take - either continue to Rendering, or report Error
+                handleActionInternal(presentContainer(actionProcessor))
+            }
 
-                is AwaitEffect -> {
-                    sideEffect.completion.await()
-                }
+            is AwaitEffect -> {
+                await()
+                // recursive call to self with the next effect
+                effect.launch()
+            }
 
-                is ProcessActions -> {
-                    actionProcessor.process(sideEffect.actions)
-                    Success(_state)
-                }
+            is ProcessActions -> {
+                actionProcessor.process(actions)
+                Success(_state)
             }
-        } else {
             // if no side effect, return success with current state
-            Success(_state)
-        }.also {
-            if (_state == Idling) {
-                stopLifecycleTracking()
-            }
+            else -> Success(_state)
         }
     }
 
-    private fun State.take(action: Action): Transition {
+    private fun State.takeTransition(action: Action): Transition {
```

Comment:
I prefer the previous naming, as the `State` is taking the `Action`, and producing the `Transition`

Reply:
forgot to take out before pushing it